src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -808,28 +829,37 @@ private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<
                 hasError = true;
             }
 
+            // If this is an UnmanagedCallersOnly method, it means that this cannot be called by managed code, including the attempt by the CLR
+            // to run the module initializer.
+            if (_lazyCustomAttributesBag.EarlyDecodedWellKnownAttributeData is CommonMethodEarlyWellKnownAttributeData { UnmanagedCallersOnlyAttributeData: { } methodData })
+            {
+                Debug.Assert(!ReferenceEquals(methodData, UnmanagedCallersOnlyAttributeData.Uninitialized));
+                arguments.Diagnostics.Add(ErrorCode.ERR_ModuleInitializerCannotBeUnmanagedCallersOnly, arguments.AttributeSyntaxOpt.Location);
+                hasError = true;
+            }
+
             if (!hasError && !CallsAreOmitted(arguments.AttributeSyntaxOpt.SyntaxTree))
             {
                 DeclaringCompilation.AddModuleInitializerMethod(this);
             }
         }
 
-        private void DecodeUnmanagedCallersOnlyAttribute(DecodeWellKnownAttributeArguments<AttributeSyntax, CSharpAttributeData, AttributeLocation> arguments)
+        private void UnmanagedCallersOnlyAttributeErrors(CSharpAttributeData attributeData, AttributeSyntax syntax, DiagnosticBag diagnostics)
         {
-            Debug.Assert(arguments.AttributeSyntaxOpt is not null);
             if (!IsStatic || MethodKind is not (MethodKind.Ordinary or MethodKind.LocalFunction))
             {
                 // `UnmanagedCallersOnly` can only be applied to ordinary static methods or local functions.
-                arguments.Diagnostics.Add(ErrorCode.ERR_UnmanagedCallersOnlyRequiresStatic, arguments.AttributeSyntaxOpt.Location);
+                diagnostics.Add(ErrorCode.ERR_UnmanagedCallersOnlyRequiresStatic, syntax.Location);
             }
 
-            if (!arguments.Attribute.CommonNamedArguments.IsDefaultOrEmpty)
+            // Early decoding cannot issue errors on invalid types being used for calling convention types, so we'll
+            // check them now.
+            if (attributeData.CommonNamedArguments is { IsDefaultOrEmpty: false } namedArgs)
             {
-                foreach (var (key, value) in arguments.Attribute.CommonNamedArguments)
+                foreach (var (key, value) in namedArgs)
                 {
-                    if (key != "CallConvs"
-                        || value.Kind != TypedConstantKind.Array
-                        || value.Values.Any(v => v.Kind != TypedConstantKind.Type))
+                    if (!UnmanagedCallersOnlyAttributeData.IsCallConvsTypedConstant(key, in value)
+                        || value.Values.IsDefaultOrEmpty)
```

Comment:
>|| value.Values.IsDefaultOrEmpty [](start = 24, length = 32)

Is this check necessary?

Reply:
I suppose not.

---
In reply to: [483169534](https://github.com/dotnet/roslyn/pull/47311#discussion_r483169534) [](ancestors = 483169534)